### PR TITLE
Representation now can be compared to check if they are equivalent. solves issue #7793

### LINKED ISSUE #7793: Absence of representation comparison makes frame.is_equivalent_frame incorrectly fail
Paraphrasing from #3182, at the moment one cannot compare whether representation objects are equivalent:

```
In [1]: from astropy.coordinates import CartesianRepresentation

In [2]: CartesianRepresentation(x=0, y=0, z=0, unit="km") == CartesianRepresentation(x=0, y=0, z=0, unit="km")
Out[2]: False

In [3]: CartesianRepresentation(x=0, y=0, z=0, unit="km") == CartesianRepresentation(x=1000, y=1000, z=1000, unit="km")
Out[3]: False
```

However, this issue is arguably more important because it makes `is_equivalent_frame` fail when it shouldn't:

```
In [4]: from astropy.coordinates import GCRS, HCRS

In [5]: from astropy.time import Time

In [6]: epoch = Time.now()

In [7]: HCRS(obstime=epoch).is_equivalent_frame(HCRS(obstime=epoch))
Out[7]: True

In [8]: GCRS(obstime=epoch).is_equivalent_frame(GCRS(obstime=epoch))
Out[8]: False

In [9]: GCRS(obstime=epoch).obsgeoloc
Out[9]: 
<CartesianRepresentation (x, y, z) in m
    (0., 0., 0.)>

In [10]: GCRS(obstime=epoch).obsgeoloc == GCRS(obstime=epoch).obsgeoloc
Out[10]: False
```

### PR DESCRIPTION
**Pull request by souravghosh97**
Previously comparing equivalent frame may lead to wrong results since one cannot compare representation objects are equivalent. 
Previously 
```
>>> from astropy.coordinates import GCRS
>>> from astropy.time import Time
>>> epoch = Time.now()
>>> GCRS(obstime=epoch).is_equivalent_frame(GCRS(obstime=epoch))
>>> False
```
But now

```
>>> GCRS(obstime=epoch).is_equivalent_frame(GCRS(obstime=epoch))
True
```

EDIT: Fix #7793

### CODE REVIEW

**Comment by pllim** [comment]
cc @Juanlu001

File: astropy/coordinates/representation.py [hunk 1]
```diff
@@ -324,6 +324,17 @@ def __sub__(self, other):
...
         return self._combine_operation(operator.sub, other, reverse=True)
 
+    # Required to check equality between representation and differential
+    # classes
+    def __eq__(self, other):
+        if self.__class__ != other.__class__:
+            return False
+
+        for component in self.components:
+            if getattr(self, component) != getattr(other, component):
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: This will not work for non-scalar components, since then you get an array of bool

**Comment by mhvk** [comment]
I feel we've had this discussion before and decided that because of rounding issues we would not allow comparisons between representations at all (@taldcroft?); the question thus is whether we can make the frame equivalency be more aware of these issues, and check for proximity rather than strict equality.

**Comment by mhvk** [comment]
p.s. @souravghosh97 - sorry for the brevity; thanks for getting the ball rolling on this!

**Comment by Juanlu001** [comment]
@mhvk What about forgetting about rounding issues and check for equality to allow simple cases like `GCRS(obstime=epoch).is_equivalent_frame(GCRS(obstime=epoch))`?

**Comment by mhvk** [comment]
@Juanlu001 - fair point, but then I do think that for representations `__eq__` has to work like `ndarray` and return an array of bool. A question then is what to do with representations that are not the same class; my tendency would be to compare as `Cartesian`, but then rounding errors certainly are a problem.

Overall, I would add that I don't think adding `__eq__` is the right way to solve the equivalency test for frames; it may work for simple cases, but then it will be all the more confusing if it doesn't work for more complicated ones... Hence, I would suggest to think through clearly what exact behaviour we want for `is_equivalent` and then implement that behaviour inside that method. When we're happy with that, we can as a second step consider whether whatever we implemented is suited to become a method on the representation classes (whether that be `__eq__` or an `is_close` or something else altogether).

File: astropy/coordinates/representation.py [hunk 2]
```diff
@@ -325,14 +325,21 @@ def __rsub__(self, other):
         return self._combine_operation(operator.sub, other, reverse=True)
 
     # Required to check equality between representation and differential
-    # classes
+    # classes.
     def __eq__(self, other):
         if self.__class__ != other.__class__:
             return False
 
+        if self.isscalar and other.isscalar:
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: @souravghosh97 - this is very long-winded: one would do `np.all(<comp1> == <comp2>)`

But as noted in the main discussion, I don't think this is the right behaviour for representations - if we implement this at all, they should return arrays of bool, i.e., something like
```
result = True
for compoment in self.components:
    result = np.logical_and(result, getattr(self, component) == getattr(other, component))
```

But I would suggest waiting with making further changes until we have reached some consensus; just returning `False` if the classes are not the same feels wrong to me too. Plus, this doesn't check for derivatives.

**Comment by mhvk** [comment]
p.s. As noted in in-line, another issue that this doesn't address is derivatives that can be attached to a representation. Surely, something at the same location but with a very different velocity shouldn't evaluate as equal. I think this again suggests that adding `__eq__` to representations is the wrong path to improve `is_equivalent` for frames.

File: astropy/coordinates/representation.py (hunk 2 above)
Comment (souravghosh97) [reply astropy/coordinates/representation.py]: Yupp, it is very hard to compare representation of two different classes due to rounding error. In that case we have to ignore very slight error. I at first tried to implement by convering both self and other to cartesian.

**Comment by bsipocz** [comment]
Superseded by #8218 that fixed the original issue.

**Pull request closed without merging**